File: Solutions/cpp/quick_sort.cpp
#include <bits/stdc++.h>
using namespace std;


//严蔚敏《数据结构》标准分割函数
int Paritition1(int A[], int low, int high)
{
  int pivot = A[low];
  while (low < high)
  {
    while (low < high && A[high] >= pivot)
    {
      --high;
    }
    A[low] = A[high];
    while (low < high && A[low] <= pivot)
    {
      ++low;
    }
    A[high] = A[low];
  }
  A[low] = pivot;
  return low;
}

void QuickSort(int A[], int low, int high) //快排母函数
{
  if (low < high)
  {
    int pivot = Paritition1(A, low, high);
    QuickSort(A, low, pivot - 1);
    QuickSort(A, pivot + 1, high);
  }
}

int main()
{
  int A[10]{1,9,2,3,4,6,6,6,7,8}; 
  QuickSort(A, 0, 9);
  for(int i=0; i<10; i++)
  {
    cout << A[i] << endl;
  }
}

File: Offer_exam/20220801 Baidu/quick_sort.py
def partition(nums, l, r):
    # this pivot is a value
    # print(l, r)
    pivot = nums[l] + 0
    while l < r:
        while l < r and nums[r] >= pivot:
            r -= 1
        nums[l] = nums[r]
        while l < r and nums[l] <= pivot:
            l += 1
        nums[r] = nums[l]
    nums[l] = pivot
    return l

def quick_sort(nums, l, r):
    if r - l < 1:
        return 
    # this pivot is an index
    pivot = partition(nums, l, r)
    quick_sort(nums, l, pivot-1)
    quick_sort(nums, pivot+1, r)    


nums = [234,15,6116,12465,7174]
quick_sort(nums, 0, len(nums)-1)
print(nums)